chainercv/links/model/resnet/resblock.py

```diff
@@ -0,0 +1,107 @@
+            bottleneck with a shortcut. If :obj:`True`, apply strided
+            convolution with the first convolution layer.
+            Otherwise, apply strided convolution with the
+            second convolution layer.
+
+    """
+
+    def __init__(self, n_layer, in_channels, mid_channels,
+                 out_channels, stride, initialW=None, stride_first=False):
+        super(ResBlock, self).__init__()
+        with self.init_scope():
+            self.a = Bottleneck(
+                in_channels, mid_channels, out_channels, stride,
+                initialW, conv_shortcut=True, stride_first=stride_first)
+            self._forward = ['a']
+            for i in range(n_layer - 1):
+                name = 'b{}'.format(i + 1)
+                bottleneck = Bottleneck(
+                    out_channels, mid_channels, out_channels, stride=1,
+                    initialW=initialW, conv_shortcut=False)
+                self.add_link(name, bottleneck)
+                self._forward.append(name)
+
+    def __call__(self, x):
+        for name in self._forward:
+            l = getattr(self, name)
+            x = l(x)
+        return x
+
+
+class Bottleneck(chainer.Chain):
+
+    """A bottleneck layer.
+
+    Args:
+        in_channels (int): The number of channels of input arrays.
+        mid_channels (int): The number of channels of intermediate arrays.
+        out_channels (int): The number of channels of output arrays.
+        stride (int or tuple of ints): Stride of filter application.
+        initialW (4-D array): Initial weight value used in
+            the convolutional layers.
+        conv_shortcut (bool): If :obj:`True`, apply a 1x1 convolution
+            to the residual.
+        stride_first (bool): If :obj:`True`, apply strided convolution
+            with the first convolution layer. Otherwise, apply
+            strided convolution with the second convolution layer.
+
+    """
+
+    def __init__(self, in_channels, mid_channels, out_channels,
+                 stride=1, initialW=None, conv_shortcut=False,
+                 stride_first=False):
+        if stride_first:
+            first_stride = stride
+            second_stride = 1
+        else:
+            first_stride = 1
+            second_stride = stride
+        super(Bottleneck, self).__init__()
+        with self.init_scope():
+            self.conv1 = Conv2DBNActiv(in_channels, mid_channels, 1,
+                                       first_stride, 0, initialW=initialW,
+                                       nobias=True)
+            self.conv2 = Conv2DBNActiv(mid_channels, mid_channels, 3,
+                                       second_stride, 1, initialW=initialW,
+                                       nobias=True)
+            self.conv3 = Conv2DBNActiv(mid_channels, out_channels, 1, 1, 0,
+                                       initialW=initialW, nobias=True,
+                                       activ=lambda x: x)
```

Comment:
Calling `Caonv2DBActiv` without activ looks tricky. How about adding `Conv2DBN`? (It is ok to make it a private class)

Reply:
Hmm...
I think the current version is fine, but alternatively I can explicitly use `conv` and `bn`.

I do not like the idea of adding a private class. This usages come up a lot.
I am less hesitant on the idea of adding `Conv2DBN`, but personally it looks redundant given that this is just a `Conv2DBNActiv` with no activation....

Reply:
If you don't like adding a private class, using `Conv2D` + `BatchNormalization` is better.
Another solution is adding `active='no'` option. to `Conv2DBNActiv`.

Reply:
Since the default value of `active` is `chainer.functions.relu` (https://github.com/chainer/chainercv/blob/master/chainercv/links/connection/conv_2d_bn_activ.py#L65), we can use `active=None` for no activation.

Reply:
OK

Reply:
updated